invenio_rdm_records/records/models.py

```diff
@@ -54,6 +56,11 @@ class RDMRecordMetadata(db.Model, RecordMetadataBase, ParentRecordMixin):
     media_bucket_id = db.Column(UUIDType, db.ForeignKey(Bucket.id))
     media_bucket = db.relationship(Bucket, foreign_keys=[media_bucket_id])
 
+    # The deletion status is stored in the model so that we can use it in SQL queries
+    deletion_status = db.Column(
+        db.String(1), nullable=False, default=RecordDeletionStatusEnum.PUBLISHED.value
```

Comment:
I can't make general statements here, but in this particular case here I just didn't know that it existed :sweat_smile: 
At some point I should really take the time to _at least_ skim through the entire documentation...

Sounds sweet though, I'm gonna give it a shot and see what happens

Reply:
From quick testing, it seems to be working just fine, thanks for the pointer!

Setting it to a valid value from the enumeration will work as expected, and setting it to something invalid won't let the transaction commit:
```python
StatementError: (builtins.ValueError) 'S' is not a valid RecordDeletionStatusEnum
[SQL: UPDATE rdm_records_metadata SET updated=%(updated)s, version_id=%(version_id)s, deletion_status=%(deletion_status)s WHERE rdm_records_metadata.id = %(rdm_records_metadata_id)s AND rdm_records_metadata.version_id = %(rdm_records_metadata_version_id)s]
[parameters: [{'deletion_status': 'S', 'updated': datetime.datetime(2023, 8, 8, 14, 20, 56, 946563), 'rdm_records_metadata_version_id': 29, 'version_id': 30, 'rdm_records_metadata_id': UUID('b94589e1-7d1f-4b7f-a0c3-a654d2913a5c')}]]
```

Reply:
FTR: Alembic didn't detect any changes to the DB schema after switching to the `ChoiceType` with the `String(1)` implementation.
Not that this would've been expected, but still good to check and write down.

Reply:
Actually, the [column type detection is a bit flaky behavior with Alembic](https://alembic.sqlalchemy.org/en/latest/autogenerate.html#affecting-the-rendering-of-types-themselves), and sometimes you have to make the conversion "by-hand".

Reply:
So from what I understand there, the `repr(<column_type>)` would be relevant.
This is what I get in `invenio shell`:
```python
In [11]: repr(ChoiceType(RecordDeletionStatusEnum, impl=db.String(1)))
Out[11]: 'ChoiceType(length=1)'
```

Reply:
Update: `ChoiceType(length=1)` won't work because that's not a valid constructor call for `ChoiceType` (duh), but `String(1)` is fine.
However, testing the alembic script again did reveal an issue; namely the columns not being nullable but alembic didn't add default values.

I've just added a fix for that (by adding `server_default="P"` to the `add_column()`) to this PR.

Reply:
I think it's fine to hardcode the final DB column type in the alembic recipe, i.e. `db.String(1)`. 

I would remove the `server_default` though, since it sets the default on the column definition in the DB vs. the `default=...` we set in the Model classes field, which is Python/ORM-only. The difference is that with `server_default` if you would run an SQL INSERT statement via `psql` without passing `deletion_status` you would get the default value `"P"` in the row, which is actually not what we want: the application logic should be responsible for what values get set (so in case we want to update the default value in the future, we don't need a new alembic recipe).

Reply:
Also for the [`PersistentIdentifier.status` alembic recipe](https://github.com/inveniosoftware/invenio-pidstore/blob/d54359c185c92f87decc8b9ba951ed03790226f6/invenio_pidstore/alembic/999c62899c20_create_pidstore_tables.py#L32C57-L32C57) we're not specifying any kind of "default" either.

Reply:
hmm... yeah good point!
we're still going to need some sort of default value for the new column though, as the normal `default=...` won't update existing records, and leave them `null` (which contradicts the `not null` clause and fails the migration).
probably a simple `UPDATE` statement in the same transaction could do the trick.

Reply:
Rewrote the upgrade script, it's now a 3-step process that should work for both new instances as well as existing instances:
1. Create the new columns but make them nullable
2. Set a default value for all existing rows (s.t. they don't have null values) with a heuristic that tries to get the default value from the module
3. Set the new column to be *not nullable*